### ISSUE TITLE
Invoke-IcingaCheckCPU : Incorrect high % values

### ISSUE THREAD

**Issue by Aleksey-Maksimov** [Icinga/icinga-powershell-plugins#404]
Hello.

Sometimes Invoke-IcingaCheckCPU shows strange load % data greater than 100%

![изображение](https://github.com/Icinga/icinga-powershell-plugins/assets/17614847/0c0dd963-7788-4f42-85fb-7552eaf17fd7)

```
Environment configuration:

PowerShell Root                 => C:\Program Files\WindowsPowerShell\Modules\
Icinga for Windows Service Path => C:\Program Files\icinga-framework-service\
Icinga for Windows Service User => NT Authority\NetworkService
Icinga for Windows Service Pid  => 4544
Icinga for Windows JEA Pid      =>
Icinga Agent Path               => C:\Program Files\ICINGA2\
Icinga Agent User               => NT AUTHORITY\NetworkService
Defined Default User            => NT Authority\NetworkService
Icinga Managed User             => False
PowerShell Version              => 5.1.20348.2110
Operating System                => Microsoft Windows Server 2022 Standard
Operating System Version        => 10.0.20348
JEA Context                     =>
JEA Session File                =>
Api Check Forwarder             => True
Debug Mode                      => False

Icinga for Windows Certificate:

Issuer  => CN=Icinga CA
Subject => CN=winhost002.holding.com

List of configured background daemons on this system:

Start-IcingaWindowsRESTApi
-----------
No arguments defined

List of configured background service checks on this system:
=> https://icinga.com/docs/icinga-for-windows/latest/doc/110-Installation/06-Collect-Metrics-over-Time/

No background service checks configured

List of configured repositories on this system. The list order matches the apply order:

Icinga Stable
-----------
CloneSource  =>
Enabled      => True
LocalPath    =>
Order        => 0
RemotePath   => https://packages.icinga.com/IcingaForWindows/stable/ifw.repo.json
UseSCP       => False

Installed components on this system:

Component    Version   Available
---          ---       ---
agent        2.14.2    2.14.2
apichecks    1.2.0     1.2.0
cluster      1.3.0     1.3.0
framework    1.12.3    1.12.3
hyperv       1.3.0     1.3.0
inventory    1.2.0     1.2.0
kickstart    1.4.0
mssql        1.5.0     1.5.0
plugins      1.12.0    1.12.0
restapi      1.2.0     1.2.0
service      1.2.0     1.2.0
```

This behavior has been observed on different systems - on virtual servers with a small number of cores and on physical servers with two sockets and a large number of cores

![изображение](https://github.com/Icinga/icinga-powershell-plugins/assets/17614847/7f35e2a5-b65e-451c-91bf-75af3af1f499)

**Comment by Aleksey-Maksimov** [comment]
Almost all the time, the plugin correctly shows % values, which are similar to what we see in Windows Task Manager.
But sometimes something strange happens to the plugin and it shows an unrealistically high % value.

![изображение](https://github.com/Icinga/icinga-powershell-plugins/assets/17614847/b99785fb-43b5-454b-a1e8-7576e0585254)

**Comment by LordHepipud** [comment]
Hello,

thank you for your issue. This is actually not a bug, but a feature which was introduced with Windows 8 and beyond.

The performance counter we use since the latest version of Icinga for Windows, are using the same information as the task manager. How ever, while the task manager simply caps the CPU usage to 100%, the Icinga for Windows plugins print the correct usage.

There is a detailed docs entry available from [Microsoft](https://learn.microsoft.com/en-us/troubleshoot/windows-client/performance/cpu-usage-exceeds-100)

The short version: Systems that use Intel Turbo Boost or AMD PBO (Precision Boost Overdrive), will report the current usage different, depending if they are in the boost window and use the boost clock for completing tasks, or if they are working with the base clock.

**Comment by Aleksey-Maksimov** [comment]
The article you mentioned https://learn.microsoft.com/en-us/troubleshoot/windows-client/performance/cpu-usage-exceeds-100 says that Task Manager can show more than 100%. But we've never seen more than 100% in Task Manager. In general, our virtualization hosts are not very heavily loaded and Task Manager never provides data on high load.

No less strange is that if the service is running as "NT AUTHORITY\NetworkService" then we can see these strange large %. And if we change the service launch to run under the name “NT AUTHORITY\SYSTEM” then the situation changes and we no longer see numbers above 100%.

**Comment by Aleksey-Maksimov** [comment]
@LordHepipud, can you prevent the CPU percentage from showing more than 100% at full load?

**Comment by Aleksey-Maksimov** [comment]
I think that in the plugin it is more correct to use the `\Processor Information(*)\% Processor Time` counter instead of `\Processor Information(*)\% Processor Utility`. In this case, at peak loads we will see a clear figure of 100%. 

![изображение](https://github.com/Icinga/icinga-powershell-plugins/assets/17614847/9d4a7df0-f25b-4392-a2d0-5ee62c965d3c)

There was a discussion of a similar issue in microsoft/Windows-Dev-Performance: https://github.com/microsoft/Windows-Dev-Performance/issues/78

**Comment by LordHepipud** [comment]
Thank you for the feedback. The `% Processor Time` counter was replaced a while ago, as the Task Manager is using different performance metrics and users complained, that the load reported by Icinga for Windows did not match with the Task Manager loader.

We are currently investigating the reported numbers, but as of now it seems the performance counter library of Windows is reporting odd values. While a usage beyond 100% can happen, a 400% load or even higher is not possible.

Are you running these Windows machines as virtual machines on ESXi by any chance?

**Comment by Aleksey-Maksimov** [comment]
This issue is observed on both virtual servers and Hyper-V virtualization hosts with Windows Server 2022.

**Comment by Aleksey-Maksimov** [comment]
Hello.

When can we expect a change in this issue?
There is now a new parameter for the Invoke-IcingaCheckCPU cmdlet -Limit100Percent, but I have a feeling it does not work with the Icinga PowerShell Service.

Installed components on this system (icinga agent):
agent        2.14.4
framework    1.13.2
plugins      1.13.0
service      1.2.0


I added the **IcingaCheckCPU_Switchparameter_Limit100Percent** variable for the **-Limit100Percent** option in the command properties in Icinga Director. 

![Image](https://github.com/user-attachments/assets/7cb776f8-9f43-455c-8de8-e40aebb5aee9)

In the service template properties, I set the variable value to true.

![Image](https://github.com/user-attachments/assets/d2b4af39-0a78-4f1e-be16-1628701e74c5)

However, when the command is generated for execution on the client, for some reason the -Limit100Percent option does not get there.

![Image](https://github.com/user-attachments/assets/6e195de8-8b1c-4e49-9ec6-a66665dab6c3)

![Image](https://github.com/user-attachments/assets/6a744db4-a114-44bb-9940-977d4cd029e1)

**Comment by LordHepipud** [comment]
This issue is already fixed with v1.13.0 and the observed values of up to 127% CPU usage looks fine to me.
In general, based on modern CPU architecture, a CPU will boost up to 30% to 35% from their base clock to their boost clock.

Therefore, the representation in the result is fine.

For the not working argument: You have to import the latest basket from Icinga for Windows to add the -Limit100Percent argument.

The basket configuration will update the check command with non-configurable values. Adding this value manually will not make it available to the Icinga argument parser.
